Replace GameObject.SendMessage("Method") with event delegates

HodgsonSDAS opened the pull request:
Pull request for issue https://github.com/Microsoft/HoloToolkit-Unity/issues/224

Many [experienced Unity developers](http://jacksondunstan.com/articles/3335) would probably tell you not use [GameObject.SendMessage("String");](https://docs.unity3d.com/ScriptReference/GameObject.SendMessage.html) because [it's slow](http://answers.unity3d.com/questions/327888/is-sendmessage-slow-.html) (under the hood it uses reflection), hard to debug, and it lacks direct symbol references in the editor which makes it prone to spelling errors _because_ it's a string and not a symbol. This can also make it difficult to find everywhere the method referenced can be called and could potentially increase the amount of spaghetti code.  It may be worth while to see this updated to utilize a [delegate/Event call instead](https://unity3d.com/learn/tutorials/topics/scripting/events?playlist=17117).

## TL;DR
This will make the code easier to follow in the editor, easier to debug, and will increase performance.

## [Some arguments for using a delegate vs SendMessage](http://answers.unity3d.com/questions/185980/sendbroadcast-message-vs-delegates-and-events.html):

> SendMessage and delegates are used for similar purposes, but they work entirely differently. Delegates/Events behave exactly like they do in any other .Net program. SendMessage() is called much like Unity calls other messages like Update(). Unity reflects over all the code on the game object to see if it can invoke any messages.
> 
> As far as performance. Delegates are several times faster than SendMessage. In the order of performance:
> 
`(Least) Direct Method Call ==> Delegate/Event ==> Update() ==> SendMessage (Most Expensive)`
> As far as usage, SendMessage is like a very broad direct method call. The general paradigm is that one script specifically wants to send a message to a specific object. We're just not sure how many times is can respond or if it can respond at all. We know the intended receiver, we just don't know if it can receive the message.
> 
> With events, I usually follow the idea that the event owner doesn't know who or how many objects might respond. We are simply telling the world that something happened without too much knowledge of who will use this information.

This means if you try to trigger or respond to an event with the wrong type, you'll get a compiler error... where as if you do the wrong thing with SendMessage you'll get a runtime error that's much harder to track down.  Another major benefit of using Delegate/Event is type checking. A delegate specifies specific types that must be passed as parameters, while SendMessage just accepts an object.

msftclas commented:
Hi __@HodgsonSDAS__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <p>The agreement was validated by Microsoft and real humans are currently evaluating your PR.</p>

TTYL, MSBOT;

jwittner commented:
Kind of a huge change @HodgsonSDAS - can we break this up by feature area and make sure all the test scenes pass tests of functionality on both sides of the change?

HodgsonSDAS commented:
@jwittner Of course! I fully expected this.  But I wasn't sure exactly _which_ features we wanted to break this up into.  Any suggestions?

I also fully expect a through review period.

riverar commented on Assets/HoloToolkit/Input/Scripts/GazeManager.cs:
```diff
@@ -11,6 +10,20 @@ namespace HoloToolkit.Unity
+        /// <summary>
+        /// Occurs when gaze enters a GameObjects collider.
+        /// </summary>
+        /// <param name="g"></param>
```
Nit: Incomplete xmldoc

riverar commented on Assets/HoloToolkit/Input/Scripts/GazeManager.cs:
```diff
@@ -11,6 +10,20 @@ namespace HoloToolkit.Unity
+        /// <summary>
+        /// Occurs when gaze exits a GameObjects collider.
+        /// </summary>
+        /// <param name="g"></param>
```
Nit: Incomplete xmldoc

riverar commented on Assets/HoloToolkit/Input/Scripts/KeywordManager.cs:
```diff
@@ -19,23 +17,28 @@ namespace HoloToolkit.Unity
+        public delegate void OnKeywordRecognizedEvent(string keyword);
+        public event OnKeywordRecognizedEvent OnKeywordRecognized;
+
+        [ System.Serializable]
```
Nit: Errant space in attribute declaration

riverar commented on Assets/HoloToolkit/Input/Scripts/KeywordManager.cs:
```diff
@@ -44,37 +47,36 @@ public enum RecognizerStartBehavior { AutoStart, ManualStart };
 
         private KeywordRecognizer keywordRecognizer;
-        private Dictionary<string, UnityEvent> responses;
+        private List<string> responses = null;
```
Did you test this code? Doesn't seem responses gets initialized anywhere.

HodgsonSDAS replied:
Intentionally left uninitialized.  Responses are set in the Inspector window of the Editor.

riverar replied:
Where/How? I must have missed that!

riverar commented on Assets/HoloToolkit/Input/Scripts/KeywordManager.cs:
```diff
@@ -44,37 +47,36 @@ public enum RecognizerStartBehavior { AutoStart, ManualStart };
-
-                keywordRecognizer = new KeywordRecognizer(responses.Keys.ToArray());
-                keywordRecognizer.OnPhraseRecognized += KeywordRecognizer_OnPhraseRecognized;
+                Debug.LogError( "Must have at least one keyword specified in the Inspector on " + gameObject.name + "." );
```
Nit: Use string.Format or an interpolated strings

HodgsonSDAS replied:
Agreed.  I had just kept what was already there. I'll update accordingly

riverar commented on Assets/HoloToolkit/Input/Scripts/Interactable.cs:
```diff
@@ -0,0 +1,85 @@
+namespace HoloToolkit.Unity
+{
+    /// <summary>
+    /// Base class Interactable GameObjects can inherit from
```
Nit: Useless comment, expand upon or delete

HodgsonSDAS replied:
I thought about expanding this a bit, but I wanted to discuss this a bit more with the community.

riverar commented on Assets/HoloToolkit/Input/Scripts/Interactable.cs:
```diff
@@ -0,0 +1,85 @@
+        /// <summary>
+        /// Called when a user has tapped any gameObject
+        /// </summary>
+        /// <param name="go"></param>
```
Nit: Incomplete xmldoc

riverar commented on Assets/HoloToolkit/Input/Scripts/Interactable.cs:
```diff
@@ -0,0 +1,85 @@
+        {
+            if (go == gameObject)
+            {
+                //Do something if our gameObject has been tapped
```
Nit: Space after //

riverar commented on Assets/HoloToolkit/Input/Scripts/Interactable.cs:
```diff
@@ -0,0 +1,85 @@
+            }
+            else
+            {
+                //Do something if we've tapped any object except for our own
```
Nit: Space after //

riverar commented on Assets/HoloToolkit/Input/Scripts/Interactable.cs:
```diff
@@ -0,0 +1,85 @@
+        /// <summary>
+        /// Called when a user's gaze enters any gameObject
+        /// </summary>
+        /// <param name="go"></param>
```
Nit: Incomplete xmldoc

riverar commented on Assets/HoloToolkit/Input/Scripts/Interactable.cs:
```diff
@@ -0,0 +1,85 @@
+        /// <param name="go"></param>
+        protected virtual void OnGazeEnter(GameObject go)
+        {
+            if( go == gameObject )
```
Nit: Spacing and parenthesis placement
Use preferred "If (expression)" style

riverar commented on Assets/HoloToolkit/Input/Scripts/Interactable.cs:
```diff
@@ -0,0 +1,85 @@
+        /// <summary>
+        /// Called when a user's gaze exits any gameObject
+        /// </summary>
+        /// <param name="go"></param>
```
Nit: Incomplete xmldoc

riverar commented on Assets/HoloToolkit/Input/Scripts/Interactable.cs:
```diff
@@ -0,0 +1,85 @@
+        /// <param name="go"></param>
+        protected virtual void OnGazeExit(GameObject go)
+        {
+            if( go == gameObject )
```
Nit: Ditto on if style here.

riverar commented on Assets/HoloToolkit/Input/Scripts/Interactable.cs:
```diff
@@ -0,0 +1,85 @@
+
+        protected virtual void KeywordRecognized(string keyword)
+        {
+            if ( keyword.Equals( "String" ) )
```
???

HodgsonSDAS replied:
I assume you mean the formatting here as well? 

Sorry about all of these, it's difficult when your companies formatting rules differ from standard default haha.  I pretty much had to do all the formatting by hand. (Although isn't there a way to enforce this in the project settings in VS? Not sure)

riverar replied:
Classes that don't override the method end up doing some unnecessary work. I like your idea of providing examples though, maybe we could just comment out the code? Or do we want to just make this a documentation thing?

HodgsonSDAS replied:
I like both.

stbertou replied:
I reckon your base class like Interactable here should not have any useless "example" code but indeed you should have a InteractableExample class in the Unity-Examples folder that explains how this works.

riverar commented on Assets/HoloToolkit/Input/Scripts/Interactable.cs:
```diff
@@ -0,0 +1,85 @@
+        /// Called when a user has tapped any gameObject
+        /// </summary>
+        /// <param name="go"></param>
+        protected virtual void OnTap(GameObject go)
```
Should we just mark these unimplemented methods as abstract?

HodgsonSDAS replied:
I thought about that, but I wanted to include the examples of usage in the methods to help new contributors understand how to use and implement in their own classes.

riverar commented on Assets/HoloToolkit/Input/Tests/Scripts/ChangeColorOnFocus.cs:
```diff
@@ -0,0 +1,43 @@
+
+    protected override void OnGazeEnter(GameObject go)
+    {
+        //Check to make sure we're the game object our gaze has entered
```
Nit: Space after //

riverar commented on Assets/HoloToolkit/Input/Tests/Scripts/ChangeColorOnFocus.cs:
```diff
@@ -0,0 +1,43 @@
+
+    protected override void OnGazeExit(GameObject go)
+    {
+        //Check to make sure we're the game object our gaze has left
```
Nit: Space after //

riverar commented on Assets/HoloToolkit/Input/Tests/Scripts/ChangeColorOnSelect.cs:
```diff
@@ -0,0 +1,39 @@
+
+    protected override void OnTap(GameObject go)
+    {
+        //Check to make sure this is the game object we've tapped
```
Nit: Space after //

riverar commented on Assets/HoloToolkit/Input/Tests/Scripts/ChangeColorOnSelect.cs:
```diff
@@ -0,0 +1,39 @@
+    protected override void OnTap(GameObject go)
+    {
+        //Check to make sure this is the game object we've tapped
+        if( go == gameObject)
```
Nit: If style (see earlier comments)

riverar commented on Assets/HoloToolkit/Input/Tests/Scripts/SphereKeywords.cs:
```diff
@@ -1,13 +1,14 @@
+    protected override void KeywordRecognized(string keyword)
     {
-        switch(color.ToLower())
+        switch(keyword.ToLower())
```
We'll need to eventually rethink this design and remove the switch and .ToLower() use due to the limiting ordinal comparison going on behind the scenes (diacritics come to mind) and the tiny perf hit. OK for now.

HodgsonSDAS replied:
Agreed.  This is mainly just in a test scene, but I do use keyword.ToLower() as well in the KeywordManager

riverar commented on Assets/HoloToolkit/SpatialMapping/Scripts/TapToPlace.cs:
```diff
@@ -26,25 +26,35 @@ public partial class TapToPlace : MonoBehaviour
         /// <summary>
         /// Called when the local anchor store is ready.
         /// </summary>
         /// <param name="store"></param>
```
Nit: Incomplete xmldoc

riverar commented on Assets/HoloToolkit/Utilities/Tests/Scripts/GestureResponder.cs:
```diff
@@ -1,11 +1,31 @@
-﻿using UnityEngine;
-using System.Collections;
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
```
Nit: Do we know for sure this is copyright by Microsoft? You're not an employee so can't assert that.

HodgsonSDAS replied:
I believe that any contributions to the HoloToolkit requires you to add this, regardless of employment status. @NeerajW could you clarify this for us?

riverar replied:
The CLA just establishes a _license_ for Microsoft (and others) to use/distribute/etc. your submitted work. They do not require transfer of copyright/ownership -- you keep that!

> a. Copyright License. You grant Microsoft, and those who receive the Submission directly or indirectly from Microsoft, a perpetual, worldwide, non-exclusive, royalty-free, irrevocable license in the Submission to reproduce, prepare derivative works of, publicly display, publicly perform, and distribute the Submission and such derivative works, and to sublicense any or all of the foregoing rights to third parties.

In this case, I just wasn't sure who owned the file. The owner, or Microsoft, should be the one to document the owner. But you're probably 100% right that this piece of code is owned by Microsoft.

riverar commented on Assets/HoloToolkit/Utilities/Tests/Scripts/GestureResponder.cs:
```diff
@@ -1,11 +1,31 @@
+    private void OnTap(GameObject go)
     {
-        PlaneTargetGroupPicker.Instance.PickNewTarget();
+        //Check to make sure we've tapped on the object this script is attached to
```
Nit: Space after //

HodgsonSDAS commented:
Thanks @riverar for your feedback!

stbertou commented on Assets/HoloToolkit/Input/Scripts/OnGazeEnterEvent.cs:
```diff
@@ -1,21 +0,0 @@
-﻿using UnityEngine;
```
Just wondering why did you delete the *Event files ?
Interactable doesn't give you the option to setup Unity Event straight from the inspector (unless I'm missing something) and that was all the point of those files..

HodgsonSDAS replied:
The desire to let designers use the Inspector to set up Events this way is great for doing quick prototyping, but isn't necessarily the best design practice.  I've already discussed separating this out as a new PR and trying to get that same functionality (assigning events in Inspector) without using Unity Events. 

For reference:
http://jacksondunstan.com/articles/3335

HodgsonSDAS replied:
@stbertou Just so you know, these changes will most likely not make it through the PR, but it is something I'm looking at improving in the future.

stbertou replied:
Please keep those files but just make them use the new design pattern.
"The desire to let designers use the Inspector to set up Events this way is great for doing quick prototyping, but isn't necessarily the best practice in a production environment" this is your opinion, mine is the exact opposite hence why I added those files ;-)
So I guess as long as they don't rely on SendMessage anymore there's no reason to remove them.
Thanks!

HodgsonSDAS replied:
True, this is an opinion of mine, but if i could just do something in the Editor, I'd much prefer it to coding.  But if doing something in the Editor effects the performance of my application vs coding it, I'll take the coding over editor ease of use.

stbertou replied:
Bare in mind that some people are happy to use the inspector but they're not coders so for simple things being able to implement a behaviour on gaze events straight from the editor is extremely useful and powerful !
For just gaze events, the performance hit is really negligible.
Anyway removing those files isn't related to your change, you can keep the current scripts by just updating them to the new design pattern (or even don't touch them as the old SendMessage way would have to stick around for a while)

HodgsonSDAS replied:
Haha, that's actually how I started out in Unity. I was just a designer playing with values in the inspector!  Then I got curious... too curious haha!

btw check this out: 
https://forum.unity3d.com/threads/custom-ui-event-system-revamped.353544/

Looks pretty cool. Open source as well.

jwittner commented:
I'd really prefer if we layered support for these events on top of the existing SendMessage functionality. This change as it stands pretty much breaks all my apps and would introduce a large swath of required changes if I wanted to update to the latest HTK post it's inclusion.

HodgsonSDAS commented:
> I'd really prefer if we layered support for these events on top of the existing SendMessage functionality. This change as it stands pretty much breaks all my apps and would introduce a large swath of required changes if I wanted to update to the latest HTK post it's inclusion.

I understand, and I've addressed these concerns in the comments of issue https://github.com/Microsoft/HoloToolkit-Unity/issues/224

riverar commented:
If you guys want to keep SendMessage and sunset it, we'll need a clear deprecation plan and mark the methods as such. I prefer we gut it now and inconvenience the small user base. There are only 77 apps in the Store to date, most of which aren't real apps.

HodgsonSDAS commented:
In fact, if we can remove SendMessage from the Test scenes and only keep it around just for people we are still using it, I think this might be the easiest way to ensure we don't step on toes, while keeping the examples and test scenes clean.

NeerajW commented on Assets/HoloToolkit/Input/Scripts/Interactable.cs:
```diff
@@ -0,0 +1,89 @@
+namespace HoloToolkit.Unity
+{
+    /// <summary>
+    /// Base class Interactable GameObjects can inherit from.
```
Very vague description. Please make it more specific.
I intend to use Doxygen to generate API docs for the toolkit.

HodgsonSDAS replied:
Intentionally kept it a bit vague so we could update it later as more people commented on the nature of the changes requested.

NeerajW commented on Assets/HoloToolkit/Input/Scripts/Interactable.cs:
```diff
@@ -0,0 +1,89 @@
+        /// <param name="keyword">Spoke Keyword.</param>
+        protected virtual void KeywordRecognized(string keyword)
+        {
+            if (keyword.Equals("string"))
```
Recommend instead of "string" something like "replace with your keyword".

HodgsonSDAS replied:
Great Idea!

stbertou replied:
I still think this function should be empty, maybe just have a "do nothing" comment or an explanation.
If I don't override this method, I don't expect the base class to do anything.

NeerajW commented on Assets/HoloToolkit/Input/Scripts/KeywordManager.cs:
```diff
@@ -2,79 +2,97 @@
 {
     /// <summary>
-    /// KeywordManager allows you to specify keywords and methods in the Unity
+    /// KeywordManager allows you to specify keywords in the Unity
```
This removes the ability to hook keyword functions from the editor.
@keveleigh thoughts?

HodgsonSDAS replied:
Does not remove the ability to hook keyword functions, just changes the implementation of how to do it.

NeerajW commented on Assets/HoloToolkit/Input/Tests/Scripts/ChangeColorOnFocus.cs:
```diff
@@ -0,0 +1,43 @@
+/// <summary>
+/// Controls object appearance by changing its color when focused.
+/// </summary>
+public class ChangeColorOnFocus : Interactable
```
This feels like a test script showing usage. Any reasons for this it's in the Scripts folder?

HodgsonSDAS replied:
I agree, it should be moved into the test/scripts folder

HodgsonSDAS replied:
This is part of the test scenes.

NeerajW commented on Assets/HoloToolkit/Input/Tests/Scripts/ChangeColorOnFocus.cs:
```diff
@@ -0,0 +1,43 @@
+
+    private void Start()
+    {
+        materialInstance = GetComponent<Renderer>().material;
```
I encourage not using this pattern.
1. renderer.material.color will make a copy of the material in memory increasing memory and draw calls
2. renderer.sharedmaterial.color is better than 1 but will change all objects that use that material
3. Best is you use a shader on the material which exposes a highlight color and use this script to change that color. To reference this code, please look at the Holograms 210 Gaze academy code for highlighting objects. It uses a custom shader for this which is more performant than 1 and 2.

HodgsonSDAS replied:
It's not unchanged from the original file FocuesedObjectReciever.cs
```
 -    private void Start()
 -    {
 -        material = GetComponent<Renderer>().material;
 -        originalColor = material.color;
 -
}    
```

I agree, but probably won't make this change as part of this PR.

NeerajW commented on Assets/HoloToolkit/Input/Tests/Scripts/ChangeColorOnSelect.cs:
```diff
@@ -0,0 +1,39 @@
+using HoloToolkit.Unity;
+
+/// <summary>
+/// Controls object appearance by changing its color when selected
```
Again this script feels like a test scene script?

NeerajW commented on Assets/HoloToolkit/Input/Tests/Scripts/ChangeColorOnSelect.cs:
```diff
@@ -0,0 +1,39 @@
+
+    private void Start()
+    {
+        materialInstance = GetComponent<Renderer>().material;
```
Please see material comments above.

HodgsonSDAS replied:
Again, this class was renamed, and probably should be addressing in a separate PR or Issue in the main branch of the HTK.

NeerajW commented on Assets/HoloToolkit/SpatialMapping/Scripts/TapToPlace.cs:
```diff
@@ -26,25 +26,35 @@ public partial class TapToPlace : MonoBehaviour
+        private WorldAnchor savedAnchor;
 
-        bool placing = false;
+        private bool placing;
```
Reason for not initializing placing?

HodgsonSDAS replied:
Redundant initialization.  Bools are always false by default.

riverar replied:
bool default value is false

HodgsonSDAS commented on Assets/HoloToolkit/Input/Scripts/GestureManager.cs:
```diff
@@ -147,14 +157,14 @@ private void InteractionManager_SourceLost(InteractionSourceState state)
         }
 
-        private void OnTap()
+        private void SendOnTap()
```
Renamed to "ProcessTap"

stbertou replied:
How come we don't see this change, have you updated the review ?

HodgsonSDAS replied:
Yes, but I don't think older comments update properly in the comments thread.

HodgsonSDAS commented on Assets/HoloToolkit/Input/Scripts/GazeManager.cs:
```diff
@@ -100,13 +113,13 @@ private void UpdateRaycast()
-                if (oldFocusedObject != null)
+                if (oldFocusedObject != null && OnGazeExit != null)
                 {
-                    oldFocusedObject.SendMessage("OnGazeLeave", SendMessageOptions.DontRequireReceiver);
```
See latest PR commit.

stbertou replied:
IMO this kind of API change should only happen once we have proper release notes / versioning system in place. 
This is a major change and so far there's no good way to communicate this with people using the toolkit.
I think we should push this kind of changes very carefully.. those steps sound reasonable to me:
1) add the new feature but still default to the old one, update the release notes so people are aware of this new design pattern.
also let us know that in for example v 2.0 the code will be deprecated 
2) eventually iterate on 1.0 for a while adding non related changes
3) once it's time to deprecate the old system, add deprecate warnings like you did but keep the old stuff so people have time to convert their code
Obviously the release notes should be clear about this upcoming API change
4) once we move to 2.0 it's fine to remove the old code
Any thoughts?

stbertou replied:
@NeerajW @jwittner any thoughts on this ? don't think it's the perfect excuse to at least put a release notes system in place?

HodgsonSDAS commented on Assets/HoloToolkit/Input/Scripts/GazeManager.cs:
```diff
@@ -11,6 +10,20 @@ namespace HoloToolkit.Unity
+        /// Occurs when gaze enters a GameObjects collider.
+        /// </summary>
+        /// <param name="g">The GameObject you gaze has entered.</param>
+        public delegate void OnGazeEnterEvent(GameObject g);
```
Will do!

HodgsonSDAS commented on Assets/HoloToolkit/Input/Prefabs/FocusedObjectKeywordManager.prefab:
```diff
@@ -20,7 +20,6 @@ GameObject:
   m_Component:
```
I agree.  But we should address that there are currently 4 meta files that keep getting regenerated because they were not properly added in the main branch:

two .fbm/templates.meta files for the cursor
Assets/HoloToolkit/Input/Plugins.meta
Assets/HoloToolkit/Scripts/Microphone.meta

keveleigh commented on Assets/HoloToolkit/Input/Scripts/KeywordManager.cs:
```diff
@@ -84,26 +102,30 @@ void OnDestroy()
+            if (OnKeywordRecognized != null && responses.Contains(args.text))
             {
-                keywordResponse.Invoke();
+                OnKeywordRecognized(args.text);
```
It also causes me to need a specific method for handling keywords in various scripts ([like this one you mentioned above](https://github.com/HodgsonSDAS/HoloToolkit-Unity/blob/3a4549ab0092827f7539b0b77e0da72513ffc4e8/Assets/HoloToolkit/Input/Scripts/Interactable.cs#L81)). In the current implementation, I can call any public method I want as the event, allowing me to reuse code easily instead of routing it through another method.

HodgsonSDAS replied:
http://jacksondunstan.com/articles/3335

keveleigh replied:
Yeah, the reason we use UnityEvents is for Inspector exposure. I don't believe there's another way to duplicate this functionality, but I'll be happy to be proven wrong!

HodgsonSDAS replied:
I'd definitely like to see if we can do this.

HodgsonSDAS replied:
Short answer, yes it is doable. 

Long answer, who wants to write it? lol.
There are a few examples on the web and in the unity asset store.

keveleigh replied:
Agreed on the long answer haha! Since Unity gives us a simple way to do what we want with UnityEvents, my recommendation would be to leave this implementation as is, especially since these events only fire once when a keyword is recognized (instead of every frame or even every couple frames, when a performance increase might be more noticeable).

keveleigh commented on Assets/HoloToolkit/Input/Scripts/KeywordManager.cs:
```diff
@@ -84,26 +102,30 @@ void OnDestroy()
         private void ProcessKeyBindings()
         {
-            foreach (var kvp in KeywordsAndResponses)
+            for (int i = 0; i < KeywordsAndResponses.Length; i++)
```
Is this more efficient than foreach?

HodgsonSDAS replied:
Even encouraged to use for instead of foreach in [Unity Performance Considerations](https://developer.microsoft.com/en-us/windows/holographic/performance_recommendations_for_unity)

keveleigh replied:
I was under the assumption (and that article seems to agree: "Avoid foreach loops on everything except raw arrays.") that foreach on arrays is fine but to avoid foreach on lists and other collections.

HodgsonSDAS replied:
That is true, but this loop is embedded in the update loop, and we'll need to iterate this array every frame.  It might be best to keep it as a for loop to avoid any memory allocations during every frame.

This could potentially have a big impact of performance if we were to have a fairly large array, although, I think when it comes down to it, testing this and doing a study to see exactly what the performance impact would be might be best.

HodgsonSDAS replied:
That is true, but this loop is embedded in the update loop, and we'll need to iterate this array every frame. It might be best to keep it as a for loop to avoid any memory allocations during every frame.

This could potentially have a big impact of performance if we were to have a fairly large array, although, I think when it comes down to it, testing this and doing a study to see exactly what the performance impact would be might be best.

keveleigh replied:
I agree! This implementation has been on my list of things to look at since it was added to the toolkit. It would probably be worth wrapping the call in Update with
`if (Input.anyKeyDown)`
so it's only called [during the frame a key is actually pressed](https://docs.unity3d.com/ScriptReference/Input-anyKeyDown.html), that way this loop will only happen when necessary instead of every frame.

keveleigh commented on Assets/HoloToolkit/Input/Scripts/Interactable.cs:
```diff
@@ -0,0 +1,89 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
```
This is a new file and should probably have its own pull request.

HodgsonSDAS replied:
This is required class that is needed to facilitate the changes of the PR.

HodgsonSDAS commented on Assets/HoloToolkit/Input/Scripts/KeywordManager.cs:
```diff
@@ -2,79 +2,97 @@
+    /// Please, also make sure to also set your keyword responses in the Inspector window under Keywords and Responses Array.
     /// </summary>
-    public partial class KeywordManager : MonoBehaviour
+    public partial class KeywordManager : Singleton<KeywordManager>
```
Having them scattered around the scene may not be such a good idea and could make it quite hard to find. The class is also includes "Manager" in the name, which usually suggests that it's the only one of its kind.

keveleigh replied:
KeywordRecognizer has been optimized to have many of them around the scene, to facilitate turning keywords on/off when they're valid/invalid. I'm happy to rename this script if you think Manager implies only one. Its purpose is to manage setting up a keyword recognizer and make it easier by setting keywords in the Inspector, taking the code setup out of the hands of the developer.

What if I have a KeywordManager I want to turn on when I gaze at a specific GameObject? How does having this as a Singleton handle turning on and off specific keywords based on context? In the current implementation, we can place a KeywordManager on the object and turn it on in OnGazeEnter.

HodgsonSDAS replied:
Ahh, I see.

Yes, renaming it might be best if used in this way.

keveleigh replied:
The KeywordRecognizer has been written in a way to optimize turning them on and off when their keywords are actually needed. Leaving all keywords on and ignoring their events if the state isn't correct creates more work for the recognition engine (more keywords to try to differentiate and recognize against, increasing the chance of a false accept or reject) and is less efficient overall. The old model worked this way, since adding and removing keywords used to be expensive, but we worked with Unity to improve the implementation into what exists today.

keveleigh replied:
Also, the scripts placed around the scene would typically only contain keywords relevant to the GameObject they're on. It's still possible to have a high level recognizer for "global" voice commands without preventing an easy implementation of "in-context" voice commands (say, on a single object and turned on on gaze or on a Select).

HodgsonSDAS replied:
Okay, I'll be sure to keep this in mind in the future.

HodgsonSDAS commented on Assets/HoloToolkit/Input/Scripts/KeywordManager.cs:
```diff
@@ -2,79 +2,97 @@
+        /// </summary>
+        /// <param name="keyword">Keyword spoken.</param>
+        public delegate void OnKeywordRecognizedEvent(string keyword);
+        public event OnKeywordRecognizedEvent OnKeywordRecognized;
```
Could you elaborate a bit more?
What is it that you don't see in the Inspector?

keveleigh replied:
The OnKeywordRecognized event doesn't show up, so I'm not sure where I'd register this event.

HodgsonSDAS replied:
I think we already discussed finding a way to keep the functionality of assigning Events in the Inspector elsewhere.

izatar commented:
This is a comical case of [Bikeshedding](https://en.wiktionary.org/wiki/bikeshedding).
There are some real issues that need to be dealt with, and here we are doing peephole optimization.

HodgsonSDAS commented on Assets/HoloToolkit/Input/Scripts/GestureManager.cs:
```diff
@@ -74,7 +84,7 @@ public Vector3 ManipulationHandPosition
         {
             get
             {
-                Vector3 handPosition = Vector3.zero;
```
Initialization is redundant.  The value assigned is never used.

stbertou replied:
The method is called TryGetPosition so without checking its implementation it looks like it's possible to not get a "correct" position, is that correct ? I guess that's why the code was originally setting it to zero.
I guess it would be safer to check for the returned Boolean and return zero if TryGetPosition returned false.

stbertou replied:
Moving forward a PR should only contain changes related to a specific issue.
This way it's 1) easier to review changes and also 2) easier to revert commits or merge in case something goes wrong
Just my 2c...

darax commented on Assets/HoloToolkit/Input/Scripts/Interactable.cs:
```diff
@@ -0,0 +1,100 @@
+        /// <param name="tappedGameObject">GameObject user has tapped.</param>
+        protected virtual void OnTap(GameObject tappedGameObject)
+        {
+            if (tappedGameObject == gameObject)
```
Are we creating a different performance issue?  As the number of intractable controls increase the number of unnecessary events being fired also increases.  I suspect that there is a point where the cost of reflection in SendMessage and the cost of sending unnecessary events intersect.  I'm not sure how many objects that is, but...

Have we considered using an interface and having the GazeManager request that interface as the focused object changes?  I suspect that GetComponent<> is also not a trivial cost, but we would be enabled to cache the results of GetComponent<> for a GameObject so we'd only need to do it once.  With this approach we wouldn't have to require people to include code that ensures the event was meant for them.  We could also have the interface return if the operation was handled.

HodgsonSDAS replied:
No. See http://jacksondunstan.com/articles/3335

darax replied:
First, I agree 100% that the SendMessage() call is bad.  It's bad for performance since it uses reflection.  It encourages sloppy coding.   It makes scripts less portable.  Getting rid of SendMessage is a good idea.  

However there must be an intersection point N where SendMessage and signaling N events has the same performance.  As N increases SendMessage is cheaper.  As N decreases signaling the events is cheaper.  This number N exists.   I might try to discover 'N' in a simple application.  The article you linked tested up to N=5.  It isn't comparing C# events with SendMessage, so the side by side comparison in the article isn't relevant to my point (I see how it is for the keywordmanager wireup, but this is orthogonal).

If N turns out to be a relatively small number (less than 100?) then we should look at alternatives and an interface is an alternative (which we should, of course, measure the performance impact). I'm not sure that I consider 'avoid' to be the same as a 'No no'.  :)

HodgsonSDAS replied:
I'll advised ;)

darax replied:
ISWYDT :)  I'm working on getting some measurements.

darax replied:
It's worse than I would have predicted.  It appears that switching to events will be worse than SendMessage for N > 5 (N is the number of objects that would care to listen for the event). I'm going to see what the cost is of using an interface. 

some info from my test where n = 10.
Frame 3960: SendEvent current total 196.3016 ms 
Frame 3960: SendMessage current total 126.7389 ms 

and n = 100
Frame 1560: SendEvent current total 194.5788 ms 
Frame 1560: SendMessage current total 40.0294 ms

HodgsonSDAS replied:
strange I got wildly different results:

Number of Arguments, Number of Listeners, C# Event Time(ms), UnityEvent Time(ms)
0,1,243,3932
0,2,746,6068
0,3,1626,8613
0,4,2096,10055
0,5,3447,12109
0,6,3170,13985
0,7,3593,15937
0,8,4091,17831
0,9,4616,19753
0,10,5729,22745
1,1,244,5852
1,2,878,8165
1,3,1663,10930
1,4,2362,12918
1,5,3230,15817
1,6,2964,17036
1,7,4214,19493
1,8,3866,21435
1,9,4427,23656
1,10,4858,25973
2,1,250,7247
2,2,1002,9600
2,3,1687,12096
2,4,2023,14537
2,5,2425,17074
2,6,2924,19558
2,7,3579,24862
2,8,4425,24441
2,9,4394,25374
2,10,5015,28224

HodgsonSDAS replied:
I'm also inclined to also just test out Raycast object, get component, and call it's tapped method as well. 

I agree, when subscribing to lots and lots of listeners it definitely sees a performance hit, but not nearly as much as Unity events.

Although, I don't really see this being a problem for most developers, I _could_ see a problem for devs that are using lots and lots of intractable objects in the scene at once.

darax replied:
I really need to make sure that we are comparing the right things.  I'm not looking at **UnityEvents** at all.   :)  That test method you are using doesn't apply.  

Here's what I'm doing:

```
using UnityEngine;
using System.Collections.Generic;
using HoloToolkit.Unity;
using System;

public class SendEventTest : Singleton<SendEventTest> {

    public delegate void OnGazeEnterEvent(GameObject focusedObject);
    public event OnGazeEnterEvent OnGazeEnter;

    List<GameObject> spawnedObjects = new List<GameObject>();
    public GameObject listenerPrefab;
    public int numObjects = 100;

    private double totalms = 0;
    // Use this for initialization
    void Start ()
    {
        numObjects = 100;
	    for(int index=0;index<numObjects;index++)
        {
            spawnedObjects.Add(Instantiate(listenerPrefab));
        }
	}
	
	// Update is called once per frame
	void Update () {
        // send the event to a random object
        DateTime start = DateTime.Now;
        GameObject rto = spawnedObjects[(int)(spawnedObjects.Count * UnityEngine.Random.value)];
        if (OnGazeEnter != null)
        {
            OnGazeEnter(rto);
        }

        totalms += (DateTime.Now - start).TotalMilliseconds;

        if (Time.frameCount % 120 == 0)
        {
            Debug.LogFormat("Frame {0}: SendEvent current total {1} ms ",Time.frameCount, totalms);
        }
    }
}


```

and

`using UnityEngine;
using System.Collections.Generic;
using System;

public class SendMessageTest : MonoBehaviour {

    List<GameObject> spawnedObjects = new List<GameObject>();
    public GameObject listenerPrefab;
    public int numObjects = 100;

    private double totalms = 0;
    // Use this for initialization
    void Start ()
    {
        numObjects = 100;
        for (int index = 0; index < numObjects; index++)
        {
            spawnedObjects.Add(Instantiate(listenerPrefab));
        }
    }
	
	// Update is called once per frame
	void Update ()
    {
        DateTime start = DateTime.Now;
        GameObject rto = spawnedObjects[(int)(spawnedObjects.Count * UnityEngine.Random.value)];
        rto.SendMessage("OnSelect", SendMessageOptions.DontRequireReceiver);
        totalms += (DateTime.Now - start).TotalMilliseconds;

        if (Time.frameCount % 120 == 0)
        {
            Debug.LogFormat("Frame {0}: SendMessage current total {1} ms ", Time.frameCount, totalms);
        }

        if (Time.frameCount % 1200 == 0)
        {
            totalms = 0;
        }
    }
}`

the prefabs referenced in the scripts have either a script that has an OnSelect() function (I tried it without too) or a script that registers for an event handler.  I measuring how much time is being taken during the call.

HodgsonSDAS replied:
I'll take a look at this and compare notes.

darax replied:
BTW, I can make send message more expensive by making a deeper hierarchy.

HodgsonSDAS replied:
What's the rest of the setup?  You've also got prefabs and such.  I'm not sure what goes on there and what those scripts look like either.

darax replied:
My scene only has one object (aside from the Main camera and directional light)

It's this:
![image](https://cloud.githubusercontent.com/assets/15933792/18886910/d2923dae-84a6-11e6-9962-867ef863ba49.png)

the listeners all look like:
![image](https://cloud.githubusercontent.com/assets/15933792/18886929/e893f91c-84a6-11e6-9e7c-c4249db505a4.png)

Here's the scripts, I just added an interface:
[scriptszippedup.zip](https://github.com/Microsoft/HoloToolkit-Unity/files/496407/scriptszippedup.zip)

darax replied:
BTW, the interface seems pretty okay so long as the results from GetComponent<> are cached, otherwise the performance is similar to SendMessage.

Frame 1680: SendEvent current total 248.4189 ms 
Frame 1680: SendMessage current total 48.5825 ms 
Frame 1680: SendInterface current total 26.5289 ms 

oh, and without the 'tapCache' in the interface class:

Frame 1680: SendEvent current total 242.203400000001 ms 
Frame 1680: SendMessage current total 48.6214 ms 
Frame 1680: SendInterface current total 55.1236 ms

HodgsonSDAS replied:
I modified the test to include the SendMessage times.  Simply create an empty scene and put this on the camera. delete the directional light. press play.

```
using System;
using System.Diagnostics;
using UnityEngine;
using UnityEngine.Events;

public class TestScript : MonoBehaviour
{
    private const int NumReps = 10000000;
    private const int MaxListeners = 1;
    private const int MaxArgs = 3;

    [Serializable]
    private class IntEvent1 : UnityEvent<int>
    {
    }

    [Serializable]
    private class IntEvent2 : UnityEvent<int, int>
    {
    }

    private event Action csharpEv0;
    private event Action<int> csharpEv1;
    private event Action<int, int> csharpEv2;
    private UnityEvent unityEv0 = new UnityEvent();
    private IntEvent1 unityEv1 = new IntEvent1();
    private IntEvent2 unityEv2 = new IntEvent2();

    private string report;

    private void Start()
    {
        Stopwatch stopwatch = new Stopwatch();
        long[,] csharpEvTimes = new long[MaxArgs, MaxListeners];
        long[,] unityEvTimes = new long[MaxArgs, MaxListeners];
        long[,] sendMsgEvTimes = new long[MaxArgs, MaxListeners];

        for (int numListeners = 0; numListeners < MaxListeners; ++numListeners)
        {
            csharpEv0 += NoOp0;
            stopwatch.Reset();
            stopwatch.Start();
            for (int j = 0; j < NumReps; ++j)
            {
                if (csharpEv0 != null)
                {
                    csharpEv0.Invoke();
                }
            }
            csharpEvTimes[0, numListeners] = stopwatch.ElapsedMilliseconds;

            unityEv0.AddListener(NoOp0);
            stopwatch.Reset();
            stopwatch.Start();
            for (int j = 0; j < NumReps; ++j)
            {
                unityEv0.Invoke();
            }
            unityEvTimes[0, numListeners] = stopwatch.ElapsedMilliseconds;

            stopwatch.Reset();
            stopwatch.Start();
            for (int j = 0; j < NumReps; j++)
            {
                SendMessage("SendMessage", SendMessageOptions.DontRequireReceiver);
            }
            sendMsgEvTimes[0, numListeners] = stopwatch.ElapsedMilliseconds;
        }
        for (int numListeners = 0; numListeners < MaxListeners; ++numListeners)
        {
            csharpEv1 += NoOp1;
            stopwatch.Reset();
            stopwatch.Start();
            for (int j = 0; j < NumReps; ++j)
            {
                if (csharpEv1 != null)
                {
                    csharpEv1.Invoke(11);
                }
            }
            csharpEvTimes[1, numListeners] = stopwatch.ElapsedMilliseconds;

            unityEv1.AddListener(NoOp1);
            stopwatch.Reset();
            stopwatch.Start();
            for (int j = 0; j < NumReps; ++j)
            {
                unityEv1.Invoke(11);
            }
            unityEvTimes[1, numListeners] = stopwatch.ElapsedMilliseconds;
        }
        for (int numListeners = 0; numListeners < MaxListeners; ++numListeners)
        {
            csharpEv2 += NoOp2;
            stopwatch.Reset();
            stopwatch.Start();
            for (int j = 0; j < NumReps; ++j)
            {
                if (csharpEv2 != null)
                {
                    csharpEv2.Invoke(11, 22);
                }
            }
            csharpEvTimes[2, numListeners] = stopwatch.ElapsedMilliseconds;

            unityEv2.AddListener(NoOp2);
            stopwatch.Reset();
            stopwatch.Start();
            for (int j = 0; j < NumReps; ++j)
            {
                unityEv2.Invoke(11, 22);
            }
            unityEvTimes[2, numListeners] = stopwatch.ElapsedMilliseconds;
        }

        report = "Num Args|Num Listeners|C# Event Time (ms)|UnityEvent Time(ms)|SendMessage Time(ms)\n------------ | ------------- | ------------- | ------------- | -------------\n";
        for (int i = 0; i < MaxArgs; ++i)
        {
            for (int j = 0; j < MaxListeners; ++j)
            {
                report += i + "|" + (j + 1) + "|" + csharpEvTimes[i, j] + "|" + unityEvTimes[i, j] + "|" + sendMsgEvTimes[i, j] + "\n";
            }
        }
    }

    private void OnGUI()
    {
        GUI.TextArea(new Rect(0, 0, Screen.width, Screen.height), report);
    }

    private static void NoOp0() { }
    private static void NoOp1(int a) { }
    private static void NoOp2(int a, int b) { }
    private void SendMessage() { }
}
```

darax replied:
Okay, I ran this test.  There are flaws in the testing approach.

First, and critical:
SendMessage("SendMessage") should be changed to SendMessage("SendAMessage") or something else that doesn't resolve to something that requires parameters.  Most of the time in SendMessage as the script is written is spent in a Debug.LogError that SendMessage requires three parameters.  

After this:
![image](https://cloud.githubusercontent.com/assets/15933792/18889251/093be486-84b1-11e6-96e7-e740a99f5394.png)


                



Second:
MaxListeners should not be 1.  It should be set to some number or UI controls that might be in the scene.  Maybe 20 is a more reasonable starting point (The shell has 13 buttons on the start page.  Add a few windows/holograms to the world and 20 is not an excessive number.  My personal app has closer to 50 individual buttons in the world).  I set the MaxListeners to 20. Now this adds pressure to how long the script takes to run (it seems to hang actually), so I decreased numreps to 100000.  Now I get:

![image](https://cloud.githubusercontent.com/assets/15933792/18889383/9774e752-84b1-11e6-9f32-ebd2accf6685.png)

HodgsonSDAS replied:
A couple of things.

Why is the time constantly increasing in your test?  Shouldn't we only be counting the time for each frame, not for the aggregate of all the frames since we started?

Keep in mind that realistically none or most of these events messages will not be called _every_ frame.

I noticed you're starting the timer before figuring out which random object to send it to. This may also skew your results.

darax replied:
The time could be reset when it prints.  It's useful to accumulate multiple samples to reduce the impact of outliers.  It's similar to the test you provided, but my 'reps' are done in successive frames rather than in a loop, so I can't use one sample of time.

True, the events won't be called every frame.  They won't be called ten million times, either.  :) It's just a useful way to get a lot of samples in a short amount of time.

The timer could be moved to after the random value.  All of the classes have the same behavior though, so that is unlikely to change how the results compare with one another.

HodgsonSDAS replied:
[perfTests.zip](https://github.com/Microsoft/HoloToolkit-Unity/files/496654/perfTests.zip)

Made a few changes to your test as well.

HodgsonSDAS replied:
After going though the tests, the Interface looks like it is the best when it come to perf with 100+ listeners.

I'll see what I can do about incorporating this change.

darax replied:
The script you have, all you need to do is update the 'MaxListeners' to set the number of callers who register.  FWIW, TimeSpan (the result of DateTime-DateTime) has a ticks property as well.  Using the stop watch is fine of course. :)

I looked at the changes you made to my scripts.  The signal to receive time is interesting, but for the event test the results are going to vary quite a bit, and the debug.log in the callback should make it so that by the time the last object is signaled the time is really large.  With 100 subscribers, I would expect that you get 100 lines of 'SendEvent current total ...' that get progressively longer.

HodgsonSDAS replied:
Yeah, I still don't think my implementation quite works well.  I'm still doing some adjustments to ensure we're getting the correct times. I was starting to see some strange numbers that didn't quite make sense.  I also want to get rid of the co-routine.

darax replied:
I'd like to propose that we switch from events/delegates to an interface so we can stop digging in.  I feel like I've generated enough data to have confidence that the cost of the interface is at worst the same perf as send message and provides the code readability benefits of an event.  I also feel like I've generated enough data to show that the event pattern will be overall more expensive on the CPU for most applications.  

If you continue to dig to try to find data that supports the event handler, do be sure to have your event handler be something like an Action<GameObject> and have the event handler compare the passed in GameObject to itself and take an action (like incrementing a value that you print out and the end of the program) if they do match.  Also be sure to measure the total cost of raising the event, not just the time spent for one of the event handlers to receive the message.

HodgsonSDAS replied:
Agreed.

Also a good read on the subject: http://jacksondunstan.com/articles/3043

HodgsonSDAS replied:
ohh also this was a good read too http://www.codeproject.com/Articles/468967/Interfaces-vs-Delegates

HodgsonSDAS replied:
Here is the latest version of my perfTest I modified from @darax 

[perfTests.zip](https://github.com/Microsoft/HoloToolkit-Unity/files/499020/perfTests.zip)

I think it's important to note, that even at 1000 listeners all 3 methods never take more than a millisecond.

jwittner commented:
@riverar  @HodgsonSDAS  The issue with simply gutting SendMessage is that there's no tool time way to tell that something has changed. Objects will simply stop getting their message method called and functionality will stop working, but there will be no warning or any indication of why this has happened. I feel we must go through a deprecation phase.

There may be only 77 apps on the store, but we have no way to measure the number of apps currently in development that may be leveraging the HTK right? I have zero apps on the marketplace, but at least four in development that leverage the HTK and this functionality.

HodgsonSDAS commented:
@jwittner I agree we need a depreciation path. I've already addressed this issue in the GestureManager by adding: 

```
        private void OnSelect()
         {
             ThrowDeprecatedWarning();
         }
 
         private void OnGazeLeave()
         {
             ThrowDeprecatedWarning();
         }
 
         private void OnGazeEnter()
         {
             ThrowDeprecatedWarning();
         }
 
         private void ThrowDeprecatedWarning()
         {
             // For more information see: TODO:Link to HoloToolkit Documentation.
             Debug.LogWarning("Using SendMessage is not recommended in the HoloToolkit and has been replaced by delegate events. See Documentation");
         }
```

jwittner commented:
@HodgsonSDAS Cool! A little confused on those changes - it looks like the functionality would still be broken. To deprecate we should print the warning, but still issue the messages, i.e. call SendMessage.

HodgsonSDAS commented:
@jwittner agreed, and it indeed does, in fact, call SendMessage.

HodgsonSDAS commented:
@darax [Another good link ](http://www.sebaslab.com/whats-wrong-with-sendmessage-and-broadcastmessage-and-what-to-do-about-it/)

> Emblematic example of these bad choices is the communication system built around SendMessage and BroadcastMessage. If you use them, you should just stop already!
> 
> In fact SendMessage and BroadcastMessage are so wrong mostly because they heavily rely on reflection to find the functions to call. This of course has a given impact on performance, but performance is not the issue here, the issue is instead related to the quality of the resulting code.
> 
> What can be so bad about it? First (and foremost) the use of  a string to recognize a function is way worse than using a string to recognize an Event. Just think about it: what happens if someone in the team decides to refactor the code where the calling function is defined and decides to rename the function or delete it?
> 
> I tell you what happens: the compiler will not warn the poor programmer of the mistake is doing and even worse, the code will just run as nothing happened. When the error will be found, it could be already too late.
> 
> Even worse is the fact that the calling function can be declared as private since the system uses reflection. You know what I usually do when I find a private function that is not used inside the declaring class? I just delete it, because the code is telling me: this function cannot be used outside this class and I am not using it, so why keep useless code?

riverar commented:
Can we just cut SendMessage out already? There's not enough usage out there to care about back compat. I'm surprised to still see active discussion on what is clearly inferior.

stbertou commented on Assets/HoloToolkit/Input/Scripts/GestureManager.cs:
```diff
@@ -251,5 +272,26 @@ void OnDestroy()
+
+        public void OnSelect()
+        {
+            ThrowDeprecatedWarning();
```
As explained somewhere else it's too early to add those warnings.. I don't want to have to comment those lines once I get the latest HTK (once/if this is merged)
Those warnings should be added once releaseNotes have been put in place

riverar replied:
Are you referring to the scenario where your existing code will fire off deprecation warnings? If so, fix your code.

That said, this doesn't look like deprecation. Looks like obsolescence, would recommend renaming this appropriately and/or just nixing the API and letting the compile errors surface. (With appropriate documentation of course)

HodgsonSDAS replied:
Can't really make the compile errors surface with `[obsolete]` in everyone's code, but I can at least throw the warning every time someone throws `OnSelect`, `OnGazeEnter`, and `OnGazeLeave`.

stbertou commented on Assets/HoloToolkit/Input/Scripts/GazeManager.cs:
```diff
@@ -11,34 +11,30 @@ namespace HoloToolkit.Unity
     {
         /// <summary>
-        /// Maximum gaze distance, in meters, for calculating a hit from a GameOjects Collider.
+        /// Occurs when gaze enters a GameObjects collider.
```
Could you please revert your changes here so we only see changes related to the PR (SendMessage..) ?
I guess the only diff we should see here is the OnGazeEnter & OnGazeExit events you added.
Looks like you reordered some members as well so it doesn't help the review process.
Cheers

HodgsonSDAS replied:
I was probably going to close the PR and open a new one to clean up some of this for easier review.  This was mainly for starting the discussion and letting people comment of the proposed changes.

stbertou commented on Assets/HoloToolkit/Input/Scripts/GestureManager.cs:
```diff
@@ -251,5 +275,26 @@ void OnDestroy()
+        private void ThrowObsoleteWarning()
+        {
+            // For more information see: TODO:Link to HoloToolkit Documentation.
+            Debug.LogWarning("Using SendMessage is not recommended in the HoloToolkit and has been replaced by delegate events. See Documentation");
```
Probably too soon to spam people with warnings here.

stbertou commented on Assets/HoloToolkit/Input/Tests/Scripts/ChangeColorOnFocus.cs:
```diff
@@ -0,0 +1,43 @@
+        }
+    }
+
+    protected override void OnGazeExit(GameObject go)
```
rename go?

HodgsonSDAS replied:
Will do!  Sorry I thought I caught all those guys

stbertou commented on Assets/HoloToolkit/Input/Scripts/GestureManager.cs:
```diff
@@ -12,8 +12,9 @@ namespace HoloToolkit.Unity
     /// <remarks>
+    /// Using the Gesture Manager requires a GameObject component to subscribe to the OnTap event.
     /// When a tap gesture is detected, GestureManager uses GazeManager to find the game object.
-    /// GestureManager then sends a message to that game object.
```
probably keep the previous comment here as well

stbertou commented on Assets/HoloToolkit/Input/Scripts/Interactable.cs:
```diff
@@ -0,0 +1,84 @@
+        /// <param name="tappedGameObject">GameObject user has tapped.</param>
+        protected virtual void OnTap(GameObject tappedGameObject)
+        {
+            // Do something if ANY GameObject has been tapped.
```
I really don't think all those commented lines are needed here. The documentation + your example scripts should be enough.

HodgsonSDAS replied:
I agree. These were mainly to show examples of how people may use this in the future.

HodgsonSDAS commented:
Moved to https://github.com/Microsoft/HoloToolkit-Unity/pull/244
